Create a 'report an exception' algorithm per #958

jeremyroman opened the pull request:
This algorithm directly includes the error propagation and fallback behavior, and requires callers to supply the global scope to be used, rather than magically inferring it.

Call sites within HTML are replaced, but there is more work yet to be done.

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
When editing this comment after the PR is created, check that PR-Preview doesn't overwrite your changes.
If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [ ] At least two implementers are interested (and none opposed):
   * …
   * …
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * These changes are not intended to modify existing behavior.
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * These changes are not intended to modify existing behavior.
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: n/a
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

The following call sites are changed (unchecked ones merit some extra attention):

- [x] **report validity steps**: If the element is not being rendered, an unspecified error may be reported. No browser actually seems to fire `error` events in this case, though Safari does log to the developer console.
- [x] **invoke custom element reactions**: The mysterious "associated script" trick is carried forward here, but we should clarify which global and script this in fact is.
- [x] **run a classic script**: The script and its settings object (thus global) are readily available.
- [x] **run a module script**: Ditto (though the script plumbing may only be necessary for classic scripts, since muting is the only use).
- [x] **`reportError()`**: The global it's called on is used, contrary to that not being plumbed in previously but aligned with what one would expect.
- [x] **register an import map**: The errors here are always synthesized in _parse an import map string_, when there's typically no running script. A `type="importmap"` script can be synchronously added by other script ought to be reported to that global but is still reported to the relevant global of the script element and its document, which is the same one the register algorithm is being run for.
- [x] **HostEnqueueFinalizationRegistryCleanupJob**: `finalizationRegistry.[[CleanupCallback]].[[Callback]].[[Realm]]` doesn't seem to be used in either Chrome or Firefox (contrary to how timers work), and this can differ from `finalizationRegistry.[[Realm]]` if the function passed was created in a different realm than the registry itself.
- [ ] **HostEnqueuePromiseJob**: Not currently cross-vendor compatible (left XXX for now).
- [x] **event handler attributes**: If it fails to parse, there's no JavaScript exception sitting around, so a JS SyntaxError is synthesized.
- [x] **timers**: I'm pretty confident (with experimentation) that browsers use the global scope for the realm the callback functions come from; in the case of passing strings to `setTimeout` and `setInterval`, it's always going to be the global scope that it was called on, since that's where it's evaluated.
- [x] **microtask queuing**: Similar to timers.
- [ ] **attaching a declarative shadow root**: Not confident. It seems you would expect the global object associated with the element we're parsing, but this isn't the case (browser don't actually report this).

I also need to:
- [x] add links for the previously exported anchor names (or whatever decision Anne and Domenic come to)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10404/custom-elements.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/custom-elements.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/custom-elements.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )
<a href="https://whatpr.org/html/10404/form-control-infrastructure.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/form-control-infrastructure.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/form-control-infrastructure.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )
<a href="https://whatpr.org/html/10404/imagebitmap-and-animations.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/imagebitmap-and-animations.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/imagebitmap-and-animations.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )
<a href="https://whatpr.org/html/10404/infrastructure.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/infrastructure.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )
<a href="https://whatpr.org/html/10404/nav-history-apis.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/nav-history-apis.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/nav-history-apis.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )
<a href="https://whatpr.org/html/10404/parsing.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/parsing.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/parsing.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )
<a href="https://whatpr.org/html/10404/timers-and-user-prompts.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/timers-and-user-prompts.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/timers-and-user-prompts.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )
<a href="https://whatpr.org/html/10404/webappapis.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/webappapis.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/webappapis.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )
<a href="https://whatpr.org/html/10404/workers.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">/workers.html</a>  ( <a href="https://whatpr.org/html/10404/2a39367...0b71d14/workers.html" title="Last updated on Jul 30, 2024, 3:45 AM UTC (0b71d14)">diff</a> )

jeremyroman commented:
@domenic @annevk When you have a chance, would you mind an early review to make sure this aligns with what you had in mind as the first step of #958? There's probably a couple things that deserve a little extra attention before this is totally mergeable (see checkmarks above and class="XXX" added in this PR, some of which might be resolvable before merge), but want to make sure this is at least mostly correct.

domenic commented on source:
```diff
@@ -59303,7 +59304,10 @@ interface <dfn interface>ValidityState</dfn> {
+     <span>report an exception</span> with an <span>implementation-defined</span> value for
+     <var>element</var>'s <span>relevant global object</span> and the <span>running
+     script</span>.</p> <p class="XXX">Is this actually reported to that global in browsers? Or
+     does it care about what realm/script called this.</p>
```
Yeah, good catch, this seems worth removing IMO.

domenic commented on source:
```diff
@@ -107734,99 +107754,124 @@ document.querySelector("button").addEventListener("click", bound);
+  should be added.</p>
+
+  <p>When the user agent is required to <dfn>report an exception</dfn> <var>exception</var>, which
+  may be a JavaScript value or an <span>implementation-defined</span> value, for a particular
```
Allowing an implementation-defined value here that is somehow not a JavaScript object seems bad. I'll review the call sites where that escape hatch is used to see if we can get rid of it...

jeremyroman replied:
The cases are:
1. non-parseable event handler strings (we just check whether they can be parsed, without actually throwing a real exception)
2. multiple constraint validation failures (above, which is quite hand-wavy generally and I haven't yet tried to rig up a test page to see what browsers do in practice)

domenic replied:
I think we should remove multiple constraint validation failures, per https://software.hixie.ch/utilities/js/live-dom-viewer/?saved=12866 and https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/forms/listed_element.cc;l=557;drc=7b232da0f22e8cdf555d43c52b6491baeb87f729

For non-parseable event handler strings, I think based on http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=12864 what happens is that people invoke the parser and get an actual `SyntaxError` object back. So I'd tweak it to something like

> Attempt to parse _body_ as a FunctionBody. If parsing fails or detects an early error, let _error_ be a `SyntaxError` object representing that error.

domenic commented on source:
```diff
@@ -107734,99 +107754,124 @@ document.querySelector("button").addEventListener("click", bound);
+  may be a JavaScript value or an <span>implementation-defined</span> value, for a particular
+  <span>global object</span> <var>global</var>, optional <span
+  data-x="concept-script">script</span> <var>script</var>, and optional boolean
+  <var>exposeError</var> (default true), it must run these steps:</p>
```
Booleans should default to false, per https://infra.spec.whatwg.org/#algorithm-params .

domenic replied:
Use named parameter style for _exposeError_ (and maybe the other two, especially if we go the route of making them both optional). You can see some examples of how to do that in HTML markup by searching for the following fragment:

```html
  <p>To <dfn>fire a traverse <code data-x="event-navigate">navigate</code> event</dfn> at a
  <code>Navigation</code> <var>navigation</var> given a <span>session history entry</span> <dfn
  data-x="fire-navigate-traverse-destinationSHE"><var>destinationSHE</var></dfn> and an optional
  <span>user navigation involvement</span> <dfn
  data-x="fire-navigate-traverse-userInvolvement"><var>userInvolvement</var></dfn> (default "<code
  data-x="uni-none">none</code>"):</p>
```

and its call sites

domenic commented on source:
```diff
@@ -107734,99 +107754,124 @@ document.querySelector("button").addEventListener("click", bound);
 
-    <p class="note">Returning true in an event handler cancels the event per <span>the event handler
-    processing algorithm</span>.</p>
+        <p class="note">The actual <code data-x="">Error</code> will not be available in the owner
```
```suggestion
        <p class="note">The actual <var>exception</var> value will not be available in the owner
```

domenic commented on source:
```diff
@@ -119201,51 +119270,9 @@ interface <dfn interface>SharedWorkerGlobalScope</dfn> : <span>WorkerGlobalScope
-
-  <p w-nodev>The <span>task source</span> for the task mentioned above is the <span>DOM manipulation
-  task source</span>.</p>
+  <span w-dev>will</span> <span data-x="report an exception">report</span> it for the worker's
```
I think this can now become "will" for both editions, since the report algorithm takes care of the normative requirements.

Similarly, the last sentence is probably redundant with the normative requirements.

domenic commented on source:
```diff
@@ -72946,7 +72950,16 @@ customElements.define("x-foo", class extends HTMLElement {
+        exception">report</span> it for the associated <span
+        data-x="concept-script">script</span>'s <span
+        data-x="concept-script-settings-object">settings object</span>'s <span
+        data-x="concept-settings-object-global">global object</span> and the associated <span
```
My instinct here would be to go to _element_'s relevant global object instead, hmm.

jeremyroman replied:
Was my instinct too, but I couldn't convince myself this was always equivalent (depending on what "the associated script", a term not defined, is). For example, the reaction callback could be from another realm, and that realm's global object might be what's used.

I checked several other sites to see what's used in practice, but I had not gotten to this one yet. I now have in Chrome and Firefox at least (Safari I'd have to go home or otherwise find a Mac, I think) -- they both seem to report, in practice, in the callback's realm, for callback reactions at least (see [this test page](https://excited-jealous-saturn.glitch.me/custom-element-reaction.html)), which is not _element_'s relevant global object.

For upgrade reactions ([test page](https://excited-jealous-saturn.glitch.me/custom-element-upgrade.html)), Chrome and Firefox diverge (and again, haven't checked Safari):
* if an exception is thrown during the ctor, Chrome fires the ctor's global's onerror but Firefox runs it on the script that we entered (the root, in the test page)
* the same is true of exceptions the spec requires the UA to throw (e.g., for returning an object other than the element being constructed from the ctor)

In no case is it necessarily _element_'s relevant global object, at least as currently implemented in those browsers. Of course, probably nobody relies on this extreme fringe case in the field.

domenic replied:
Fun. I guess from some point of view using callback's global object is pretty natural; we probably want to do that for all Web IDL-initiated callback calling, per the Web IDL parts of the plan in https://github.com/whatwg/html/issues/958#issuecomment-226309871. Let's settle on that?

domenic commented on source:
```diff
@@ -93696,7 +93709,7 @@ const p2 = navigation.navigate(url2).finished;</code></pre>
     data-x="dom-ErrorEvent-colno">colno</code> initialized to <span class="XXX">appropriate values
     that can be extracted from <var>error</var> and the current JavaScript stack in the same
-    underspecified way that the <span>report the exception</span> algorithm does</span>.</p>
+    underspecified way that the <span>report an exception</span> algorithm does</span>.</p>
```
Optional bonus: we could factor this out into an algorithm that all three call sites call.

domenic replied:
I was thinking just something like

> To fill out an ErrorEvent, given a JavaScript object _error_:

domenic commented on source:
```diff
@@ -105462,6 +105475,9 @@ new PaymentRequest(&hellip;); // Allowed to use
   </ul>
 
+  <p>A <span>global object</span> has an <dfn for="global object">in error reporting mode</dfn>
+  boolean, which is initially false.</p>
```
In the past we've avoided adding too many things to all global objects, instead putting them on `#environment-settings-object`. (Like "outstanding rejected promises weak set".) I'm not sure that was a great decision though, so maybe what you've done here is reasonable too...

domenic commented on source:
```diff
@@ -107552,8 +107568,10 @@ document.querySelector("button").addEventListener("click", bound);
+       <li><p><span>Report an exception</span> given by <var>evaluationStatus</var>.[[Value]] for
+       <var>script</var>'s <span data-x="concept-script-settings-object">settings object</span>'s
+       <span data-x="concept-settings-object-global">global object</span> and <var>script</var>.
+       </p></li>
```
No whitespace after `.`

domenic replied:
> Should `<var>script</var>.` all wrap to the next line?

Yep, that's the idea. If you collapse away the whitespace all into one line, that's what the rewrapper outputs.

domenic commented on source:
```diff
@@ -107552,8 +107568,10 @@ document.querySelector("button").addEventListener("click", bound);
+       <li><p><span>Report an exception</span> given by <var>evaluationStatus</var>.[[Value]] for
+       <var>script</var>'s <span data-x="concept-script-settings-object">settings object</span>'s
+       <span data-x="concept-settings-object-global">global object</span> and <var>script</var>.
+       </p></li>
```
It occurs to me that whenever you *do* have a script, you'll almost certainly want to use its global object. Maybe we should make the global object optional, and assert that either the global object or the script are provided?

jeremyroman replied:
I'm not sure whether that's true, mostly because I'm not entirely sure where these script objects are being summoned from.

domenic commented on source:
```diff
@@ -113831,8 +113892,12 @@ scheduleWork(); // queues a task to do lots of work</code></pre>
+  <span>associated realm</span>'s <span data-x="concept-realm-global">global object</span>.</p>
+
+  <p w-nodev class="note">This global object not necessarily the same as <span>this</span>, if
+  <var>callback</var> comes from another <span>realm</span>.</p>
```
How did you determine which choice of global to use here? If you're sure you're right, then can we have a tracking issue to write web platform tests to confirm? If you're not sure, we might want to leave this as an XXX box.

jeremyroman replied:
A lovely test page https://excited-jealous-saturn.glitch.me/ (and a few more pages under that glitch for other cases, like finalization registries and import maps).

domenic reviewed (commented):
Back from vacation! Thanks so much for tackling this.

I got halfway through the review. Then I had an epiphany. (Maybe some parts of the review are still useful.)

A lot of complexity and XXX boxes here and coming from passing the script around, and trying to determine which script to use. But Looking at the Chromium code, muting is not really done by passing a script around. It bottoms out in [IsSharedCrossOrigin()](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/api/api.cc;l=3157;drc=90cac1911508d3d682a67c97aa62483eb712f69a), which is a method of `v8::Message`. In other words, it is derived from the JS engine's representation of the thrown exception, _in the same way that we are already hand-waving about for line number, column number, and URL_.

Given that we have no interop on muting, and are pretty sure the current spec doesn't match our ideal (per long discussions in #958 and a few other issues like https://github.com/whatwg/html/issues/3149) why are we trying to preserve the framework of it being derived from the script?

My suggestion is that we change the spec hand-wave on the muting. We can have it mention that the user agent will probably want to use, for classic scripts, the "muted errors" boolean. (And that for module scripts, there is no need to mute errors.) But we don't need to find a way to thread through the "right" script.

We can then change the algorithm to only take an exception and a global.

jeremyroman reviewed (commented):
responded to most comments; will address editorial comments later

annevk commented on source:
```diff
@@ -111601,12 +111659,10 @@ typedef <span>OnBeforeUnloadEventHandlerNonNull</span>? <dfn typedef>OnBeforeUnl
-       data-x="concept-error-nothandled">not handled</i> after this, then the error may be reported
-       to a developer console.</p></li>
+       <li><p><span>Report an exception</span> with an <span>implementation-defined</span> value
+       for <var>settings object</var>'s <span data-x="concept-settings-object-global">global object
+       </span> and the appropriate <span data-x="concept-script">script</span>. The value reported
```
`</span>` cannot have preceding whitespace.

annevk commented on source:
```diff
@@ -113587,7 +113643,11 @@ enum <dfn enum>DOMParserSupportedType</dfn> {
-     <span>report the exception</span>.</p></li>
+     <span data-x="report an exception">report</span> it for <var>handler</var>'s
+     corresponding JavaScript object's <span>associated realm</span>'s
+     <span data-x="concept-realm-global">global object</span>.</p>
```
Here I would expect `<span` on the previous line.

annevk reviewed (commented):
Thank you for working on this. Great to see some progress on this issue. I wanted to note that "report the exception" has many callers that we cannot break: https://dontcallmedom.github.io/webdex/r.html#report%20the%20exception%40%40html%25%25dfn

Continuing to export the term is probably the bare minimum we need to do. Follow-up issues/downstream PRs might also be needed?

jeremyroman commented:
> Thank you for working on this. Great to see some progress on this issue. I wanted to note that "report the exception" has many callers that we cannot break: https://dontcallmedom.github.io/webdex/r.html#report%20the%20exception%40%40html%25%25dfn
> 
> Continuing to export the term is probably the bare minimum we need to do. Follow-up issues/downstream PRs might also be needed?

I agree. Tracking followup PRs in the original issue makes sense to me once this PR is in a shape you are happy with. I had the task of doing that as an XXX in the original PR because I wasn't sure what the right wattsi syntax is. I'm still not sure I got it right, but `<span id="..."></span>` is at least used in one other place. I guess the alternative would be to have explicit visible `<dfn>`s for both of them in a paragraph noting that other specs should migrate.

annevk commented:
Maintaining incoming ID references is no longer enough now HTML participates in Bikeshed's cross-referencing system. Once you drop a `<dfn export>` you break downstream that relies on it.

domenic commented:
My plan was to break all of those so that they would notice and update to the new algorithm, which will have new inputs and possibly be used quite differently (e.g. if we do the automatic handling of Web IDL callbacks). That feels a bit better than having them call the new algorithm with the wrong arguments silently...

annevk commented:
Hmm. It's usually pretty frustrating when a build is broken, especially when fixing it is tricky, but also just because it might block landing other PRs. (Although I understand that outside the WHATWG folks typically don't have fatal builds, that's more of a problem worth fixing than something we should rely on I think.)

jeremyroman commented:
The alternative might be to expose "stub" dfns, rewrite the uses indexed by webdex, and then convert them into mere anchors?

annevk commented:
Right yeah, I'm not opposed to that at all. I'd just like to avoid build time errors for people trying to fix something else in their specification unrelated to this.

domenic commented on source:
```diff
@@ -107734,99 +107750,133 @@ document.querySelector("button").addEventListener("click", bound);
-  position <var>line</var>:<var>col</var>, using a particular target <var>target</var>, it must run these steps, after which the error is either <dfn
-  data-x="concept-error-handled"><i>handled</i></dfn> or <dfn data-x="concept-error-nothandled"><i>not
-  handled</i></dfn>:</p>
+  <p>When the user agent is required to <dfn data-x="extract-error">extract error information</dfn>
```
Nit: use the simpler newer phrasing,

> To **extract error information** from a JavaScript value _exception_:

here and for the next algorithm. https://infra.spec.whatwg.org/#algorithm-declaration

domenic commented on source:
```diff
@@ -107734,99 +107750,133 @@ document.querySelector("button").addEventListener("click", bound);
+  is a JavaScript value, for a particular <span>global object</span> <var>global</var>, optional
+  <span data-x="concept-script">script</span> <var>script</var>, and optional boolean <dfn
+  data-x="report-exception-redactError"><var>redactError</var></dfn> (default false), it must run
+  these steps:</p>
```
redactError doesn't appear to be actually used :).

domenic commented on source:
```diff
@@ -108007,13 +108054,10 @@ dictionary <dfn dictionary>PromiseRejectionEventInit</dfn> : <span>EventInit</sp
-   data-x="impr-error-to-rethrow">error to rethrow</span> and return.</p></li>
+   not null, then <span>report an exception</span> given by <var>result</var>'s <span
+   data-x="impr-error-to-rethrow">error to rethrow</span> for <var>global</var> and return.
+   </p></li>
```
No whitespace before `</p>`

domenic commented on source:
```diff
@@ -113583,11 +113648,18 @@ enum <dfn enum>DOMParserSupportedType</dfn> {
-     value</span> set to <var>thisArg</var>. If this throws an exception, catch it, and
-     <span>report the exception</span>.</p></li>
+     <li>
+       <p>If <var>handler</var> is a <code data-x="idl-Function">Function</code>, then <span
```
Single-space indent here

domenic reviewed (commented):
This is looking great.

So we have two cases where the global object is unclear. Previously you've had good success testing those and then making a call. Do you want to do that for the last two? Or should we box them off with XXXs, to get this over the finish line?

I think the biggest question is whether to keep the scripts or not, per my https://github.com/whatwg/html/pull/10404#pullrequestreview-2132038376. It seems like there are several cases for the script:

1. The script is clear: "run a classic script" and "run a module script".
2. An API is being called directly, e.g. reportError(). Probably the [active script](https://html.spec.whatwg.org/#active-script) would work fine here.
3. An API indirectly queues a callback, e.g. requestAnimationFrame(), queueMicrotask(), promise.then(), any DOM manipulation that triggers custom elements. Probably capturing the active script at queueing time would be correct and relatively straightforward here?
4. An API that very indirectly triggers script execution, e.g. the script that originally set an `onload=""` attribute value or the script that caused the element to load; the script that inserted HTML that was parsed by declarative shadow DOM; the script that used `document.write()` to write a `<script type="importmap">`.
5. There is definitely no script. E.g. declarative shadow DOM or `<script type="importmap">` when reading from the server response.

Your work so far seems to give the clear unambiguous answer for (1) and (5). For (2)-(4) it's more mixed; sometimes you put "the appropriate script" and sometimes you put nothing. I suspect that no browsers do muting for (4), but as you note there are obvious interop differences for (2), and I suspect there are for (3) as well.

Given this, I think two paths are reasonable:

- Make the script completely handwavey for now, as I suggested in https://github.com/whatwg/html/pull/10404#pullrequestreview-2132038376. This might be more aligned with the actual architecture anyway.
- Make it clearer when we're in cases (1) vs. (5) vs. (2)-(4). E.g. make the script argument non-optional, and always pass either the actual script value for (1), null for (5), and a `<dfn>`d "the appropriate script" for (2)-(4), where the dfn is a handwave and there's a nice XXX box discussing the problem space and linking to a followup issue. (I guess the fact that (4) and (5) end up in the same call sites in many cases can make this a bit complicated.)

What do you think?

jeremyroman commented on source:
```diff
@@ -109202,8 +109253,20 @@ dictionary <dfn dictionary>PromiseRejectionEventInit</dfn> : <span>EventInit</sp
+      throwing functions before passing them on to the function passed to the Promise superclass
+      constructor. Which global is to be used then, considering that the current realm could be
+      different at each of those steps, by using a Promise constructor or Promise.prototype.then
+      from another realm?</p>
```
Checked this in Chrome, Firefox Nightly and Safari TP on the NewPromiseReactionJob path. They all do different things.

* Chrome doesn't report the error at all, unless the subclassed reject function also throws -- then the realm in which the `throw` occurred gets the event (not necessarily the realm which defined the custom reject function, or which constructed the error) -- though I don't see in the ECMAScript spec why the subclassed reject function should be implicated at all
* Firefox reports the error to the realm which defined the Promise subclass
* Safari reports no error, either as an `error` event or to the developer console

Of those I guess Firefox's behavior may make the most sense to me, but I don't see an easy way to describe it without additional plumbing, and this feels like something that probably would require consideration by actual JS runtime experts, if we cared. My impression is that promise subclassing is [pretty disfavoured these days anyway](https://github.com/tc39/proposal-rm-builtin-subclassing).

domenic replied:
OK, let's leave this as an XXX box and merge. Can you open a new issue that gives a code example and explains the results, and the question? We can link to it from the XXX box, either in this PR or in a quick followup.

jeremyroman commented on source:
```diff
@@ -128244,10 +128282,11 @@ document.body.appendChild(text);
-       <span>insert an element at the adjusted insertion location</span> with
-       <var>template</var>, and return.</p></li>
+       <var>serializable</var>, <var>delegatesFocus</var>, and "<code data-x="">named</code>". If
+       an exception is thrown, then catch it, <span data-x="report an exception">report</span> it
```
In fact, browsers don't seem to actually do this (at least for the exception type I tried -- the throwing a NotSupportedError DOMException for an invalid shadow host name).

Chrome will report the error to the console, but doesn't fire an `error` event. Firefox doesn't report it at all as far as I can tell. I can't make declarative shadow DOM work in my copy of Safari TP, so nothing happens there, either.

domenic replied:
OK. Let's go with what you have written, and open a new issue with the code example. We can tag mfreed7 who wrote this spec text to get his take.

I suspect the outcome will be keeping the spec text you've written, and treating the lack of reports as browser bugs, since this is a relatively new feature.

I don't think an XXX box is necessary here but it wouldn't hurt to add.

jeremyroman commented:
Replied first to the more specific items. On muting (which is what the script plumbing is for), I'd hoped to avoid doing a ton of research into how that is implemented in all three major engines, at least as a prerequisite to this PR. Not opposed to yanking it entirely into implementation-defined behavior if you think that the existing plumbing for it does more harm than good.

jeremyroman commented:
As suggested, switched the script used for muting to an implementation-defined one with a "should" recommendation that it usually be the running script, and removed that argument from the few call sites which supplied it.

domenic commented:
> Replied first to the more specific items. On muting (which is what the script plumbing is for), I'd hoped to avoid doing a ton of research into how that is implemented in all three major engines, at least as a prerequisite to this PR.

Agreed; we should not block on that. The two paths I was offering in https://github.com/whatwg/html/pull/10404#pullrequestreview-2173619192 were just alternate ways of deferring the hard cases to the future.

I'll open a new issue specifically about muting which we can link to and work on as a followup.

domenic reviewed (approved):
This LGTM. I'll give it a few days in case anyone else wants to take a look, and/or we want to make some final tweaks to link to followup issues.

domenic commented:
I realized "report an error" has no external references per https://dontcallmedom.github.io/webdex/r.html. I will push a commit that removes its `<dfn>` but leaves the ID intact.

jeremyroman commented:
Unsure whether this qualifies as purely editorial. If not, I think technically we need a second implementer interested (I assume @annevk on behalf of WebKit qualifies, if he supports this?).

domenic commented:
I'm happy to count this as not requiring implementer support, using the carveout in https://whatwg.org/working-mode#non-normative-issues. It's a bit more ambitious than the examples there, but you're not changing anything here that might be controversial, from what I can tell. It's a combination of large editorial improvements and a few cases either making already-interoperable behavior explicit, or making non-interoperable behavior more obvious.

I'll still give this a few days in case others want to do any more review. Although I think @annevk is on European vacation?

jeremyroman commented:
Assuming https://github.com/whatwg/html/issues/958#issuecomment-2255198074 was intended for here, so linking it.

jeremyroman commented:
Happy with your draft commit message. The relevant issues are #10526 and #10527.
